dagger/storageproofs/bls.nim

```diff
@@ -0,0 +1,390 @@
+    ", sectors/block=", $s,
+    ", sectorsize=", $sizeof(ZChar), " bytes" 
+ 
+  return n
+
+proc hashToG1(msg: string): blst_p1 =
+  ## Hash tp curve with Dagger specific domain separation
+  const dst = "DAGGER-PROOF-OF-CONCEPT"
+  result.blst_hash_to_g1(msg, dst, aug = "")
+ 
+proc hashNameI(name: openArray[byte], i: int64): blst_p1 =
+  ## Calculate unique filname and block index based hash
+  return hashToG1($name & $i)
+
+proc generateAuthenticatorNaive(i: int64, s: int64, t: TauZero, f: File, ssk: SecretKey): blst_p1 =
+  ## Naive implementation of authenticator as in the S&W paper.
+  ## With the paper's multiplicative notation:
+  ##   \sigmai=\(H(file||i)\cdot\prod{j=0}^{s-1}{uj^{m[i][j]}})^{\alpha}
+  var sum: blst_p1
+  for j in 0 ..< s:
+    var prod: blst_p1
+    prod.blst_p1_mult(t.u[j], fromBytesBE(getSector(f, i, j, s)), 255)
+    sum.blst_p1_add_or_double(sum, prod)
+
+  blst_p1_add_or_double(result, hashNameI(t.name, i), sum)
+  result.blst_p1_mult(result, ssk.key, 255)
+
+proc generateAuthenticatorOpt(i: int64, s: int64, t: TauZero, ubase: openArray[blst_scalar], f: File, ssk: SecretKey): blst_p1 =
+  ## Optimized implementation of authenticator generation
+  ## This implementation is reduces the number of scalar multiplications
+  ## from s+1 to 1+1 , using knowledge about the scalars (r_j)
+  ## used to generate u_j as u_j = g^{r_j}
+  ##
+  ## With the paper's multiplicative notation, we use:
+  ##   (H(file||i)\cdot g^{\sum{j=0}^{s-1}{r_j \cdot m[i][j]}})^{\alpha}
+  var sum: blst_fr
+  var sums: blst_scalar
+  for j in 0 ..< s:
+    var a, b, x: blst_fr
+    a.blst_fr_from_scalar(ubase[j])
+    b.blst_fr_from_scalar(fromBytesBE(getSector(f, i, j, s)))
+    x.blst_fr_mul(a, b)
+    sum.blst_fr_add(sum, x)
+  sums.blst_scalar_from_fr(sum)
+
+  result.blst_p1_from_affine(BLS12_381_G1)
+  result.blst_p1_mult(result, sums, 255)
+
+  result.blst_p1_add_or_double(result, hashNameI(t.name, i))
+  result.blst_p1_mult(result, ssk.key, 255)
+
+proc generateAuthenticator(i: int64, s: int64, t: TauZero, ubase: openArray[blst_scalar], f: File, ssk: SecretKey): blst_p1 =
+  ## Wrapper to select tag generator implementation
+
+  # let a = generateAuthenticatorNaive(i, s, t, f, ssk)
+  let b = generateAuthenticatorOpt(i, s, t, ubase, f, ssk)
+  # doAssert(a.blst_p1_is_equal(b).bool)
+  return b
+
+proc setup*(ssk: SecretKey, s:int64, filename: string): (Tau, seq[blst_p1]) =
+  ## Set up the POR scheme by generating tags and metadata
+  let file = open(filename)
+  let n = split(file, s)
+  var t = TauZero(n: n)
+
+  # generate a random name
+  for i in 0 ..< 512 :
+    t.name[i] = byte Rng.instance.rand(0xFF)
+
+  # generate the coefficient vector for combining sectors of a block: U
+  var ubase: seq[blst_scalar]
+  for i  in 0 ..< s :
+    let (u, ub) = rndP1()
+    t.u.add(u)
+    ubase.add(ub)
+   
+  #TODO: a better bytearray conversion of TauZero for the signature might be needed
+  let signature = sign(ssk.signkey, $t)
```

Comment:
I don't remember what this mean, why aren't we doing it here?

Reply:
Usage depends on roles in the system, but in essence, this is signing the metadata (name, size, vector of coefficients) at setup time, such that anyone dealing with it in later steps knows the metadata can be trusted. 
The comment is about the $t conversion, which is not a good conversion for a specification.